### PR TITLE
KAFKA-15415: On producer-batch retry, skip-backoff on a new leader

### PR DESCRIPTION
**Pull request by msn-tldr**
When producer-batch is being retried, new-leader is known for the partition Vs the leader used in last attempt, then it is worthwhile to retry immediately to this new leader. A partition-leader is considered to be newer, if the epoch has advanced.

Code change summary -

1. `ProducerBatch` now tracks latest leader-epoch on which attempts will be made, and change in leader(epoch) is tracked through method `hasLeaderChanged`.
2. `RecordAccumulator` decides if batch should immediately-retry(skip-backoff), or backoff, on detecting a leader change.
3. All classes have UTs & e2e test in `SenderTest.java`

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/PartitionInfo.java [hunk 1]
```diff
@@ -20,26 +20,35 @@
  * This is used to describe per-partition state in the MetadataResponse.
  */
 public class PartitionInfo {
+    public static final int UNKNOWN_LEADER_EPOCH = -1;
     private final String topic;
     private final int partition;
     private final Node leader;
+    private final int leaderEpoch;
```
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/common/PartitionInfo.java]: I think it's more aligned with the existing Kafka interface to use `Optional<Integer> leaderEpoch`. For example, `org.apache.kafka.clients.consumer.OffsetAndMetadata`.

File: clients/src/main/java/org/apache/kafka/common/PartitionInfo.java [hunk 2]
```diff
@@ -20,26 +20,35 @@
...
     private final Node[] replicas;
     private final Node[] inSyncReplicas;
     private final Node[] offlineReplicas;
 
     public PartitionInfo(String topic, int partition, Node leader, Node[] replicas, Node[] inSyncReplicas) {
-        this(topic, partition, leader, replicas, inSyncReplicas, new Node[0]);
+        this(topic, partition, leader, UNKNOWN_LEADER_EPOCH, replicas, inSyncReplicas, new Node[0]);
```
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/common/PartitionInfo.java]: And then `Optional.empty()` of course.

File: clients/src/main/java/org/apache/kafka/common/PartitionInfo.java [hunk 3]
```diff
@@ -60,12 +69,19 @@ public int partition() {
...
     /**
-     * The node id of the node currently acting as a leader for this partition or null if there is no leader
+     * The node currently acting as a leader for this partition or null if there is no leader
      */
     public Node leader() {
         return leader;
     }
 
+    /**
+     * The epoch of the partition's leader.
```
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/common/PartitionInfo.java]: I think it's the "The leader epoch of this partition if known".

File: clients/src/main/java/org/apache/kafka/common/Cluster.java [hunk 4]
```diff
@@ -277,6 +277,19 @@ public Node leaderFor(TopicPartition topicPartition) {
             return info.leader();
     }
 
+    /**
+     * Get the current leader's epoch for the given topic-partition.
+     * @param topicPartition
+     * @return The epoch for partition's leader, or UNKNOWN_LEADER_EPOCH if epoch unkown.
+     */
+    public int leaderEpochFor(TopicPartition topicPartition) {
```
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/common/Cluster.java]: Again, I think `Optional<Integer>` is more usual.

File: clients/src/main/java/org/apache/kafka/common/Cluster.java [hunk 5]
```diff
@@ -277,6 +277,19 @@ public Node leaderFor(TopicPartition topicPartition) {
             return info.leader();
     }
 
+    /**
+     * Get the current leader's epoch for the given topic-partition.
+     * @param topicPartition
+     * @return The epoch for partition's leader, or UNKNOWN_LEADER_EPOCH if epoch unkown.
```
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/common/Cluster.java]: Typo. "unkown" -> "unknown".

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java [hunk 6]
```diff
@@ -94,9 +100,40 @@ public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, lon
...
 
+    /*
+     * Returns whether the leader epoch has changed since the last attempt.
+     * @param latestLeaderEpoch The latest leader epoch.
+     * @return true if the leader has changed, otherwise false.
+     */
+    boolean hasLeaderChanged(int latestLeaderEpoch) {
+        boolean leaderChanged = false;
+        // Checking for leader change makes sense only from 1st retry onwards(attempt >=1).
+        log.trace("For {}, attempting to change leader, currentLeaderEpoch:{}, leaderChangedAttempts:{}, latestLeaderEpoch: {}, current Attempt: {}",
```
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: There are 4 separate calls to `attempts()` (or the equivalent of `attempts.get()`) in this method. Capture the value with a single call and use instead.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java [hunk 7]
```diff
@@ -94,9 +100,40 @@ public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, lon
         this.isSplitBatch = isSplitBatch;
         float compressionRatioEstimation = CompressionRatioEstimator.estimation(topicPartition.topic(),
                                                                                 recordsBuilder.compressionType());
+        this.currentLeaderEpoch = PartitionInfo.UNKNOWN_LEADER_EPOCH;
+        this.leaderChangedAttempts = -1;
```
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: I don't think there's any need for a special value of -1. Initializing to 0 would work fine.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java [hunk 8]
```diff
@@ -79,6 +80,11 @@ private enum FinalState { ABORTED, FAILED, SUCCEEDED }
     private boolean retry;
     private boolean reopened;
 
+    // Tracks the current-leader's epoch to which this batch would be sent, in the current to produce the batch.
+    private int currentLeaderEpoch;
+    // Tracks the attempt in which leader was changed to currentLeaderEpoch for the 1st time.
+    private int leaderChangedAttempts;
```
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: Sounds like the number of attempts to change the leader. Perhaps `attemptsWhenLeaderLastChanged` would be more descriptive.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java [hunk 9]
```diff
@@ -94,9 +100,40 @@ public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, lon
...
+            if (currentLeaderEpoch != latestLeaderEpoch) {
+                leaderChangedAttempts = attempts();
+                leaderChanged = true;
+            } else {
+                // Otherwise, it's only a leader change until the first attempt is made with this leader
+                leaderChanged = attempts.get() == leaderChangedAttempts;
+            }
+        }
+        if (leaderChanged) {
+            log.debug("For {}, leaderChanged, currentLeaderEpoch:{}, leaderChangedAttempts:{}",
```
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: I don't see why you're tracing only the previous leader epoch here. I think you're interested in the new one, or both new and old. Because the assignment follows the log line, the "current" is still previous.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java [hunk 10]
```diff
@@ -517,4 +554,14 @@ public boolean isTransactional() {
...
         return reopened;
     }
+
+    // VisibleForTesting
+    int currentLeaderEpoch() {
+        return currentLeaderEpoch;
+    }
+
+    // VisibleForTesting
+    int leaderChangedAttempts() {
```
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: `attemptsWhenLeaderLastChanged()` perhaps.

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java [hunk 11]
```diff
@@ -156,6 +157,8 @@ public class SenderTest {
     private SenderMetricsRegistry senderMetricsRegistry = null;
     private final LogContext logContext = new LogContext();
 
+    private final Logger log = logContext.logger(SenderTest.class);
```
Comment (AndrewJSchofield) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java]: I'm surprised that you added a logger when this massive test hasn't used one so far. Seems like logging is not required for working out whether the test passed or how it failed. I suggest removing it.

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java [hunk 12]
```diff
@@ -96,6 +98,8 @@ public class RecordAccumulatorTest {
     private final long maxBlockTimeMs = 1000;
     private final LogContext logContext = new LogContext();
 
+    private final Logger log = logContext.logger(RecordAccumulatorTest.class);
```
Comment (AndrewJSchofield) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java]: Probably unnecessary.

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java [hunk 13]
```diff
@@ -3146,6 +3149,97 @@ public void testInvalidTxnStateIsAnAbortableError() throws Exception {
...
+                        } else if (tp1.equals(tp)) {
+                            return 0;
+                        } else {
+                            throw new RuntimeException("unexpected tp " + tp);
+                        }
+                    }));
+            sender.runOnce(); // send produce request, immediately.
+            assertEquals(1, sender.inFlightBatches(tp0).size());
+            assertTrue(client.hasInFlightRequests());
+            client.respond(produceResponse(tp0, -1, Errors.NOT_LEADER_OR_FOLLOWER, 0));
```
Comment (AndrewJSchofield) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:3217]: Should there not be testing that uses the new `CurrentLeader` field in v9 of the ProduceResponse given that this is part of KIP-951 which this PR is implementing?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java (hunk 7 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: Seems as though 0 would do, yes.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java (hunk 6 above)
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: Can I be extremely nit-picky about the capitalization of `Attempt` in the log line? Any reason it can't just be `attempt`?

Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: Also super nit-picky: spaces after colons in message.

Sorry, I can't help myself 😛

Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: The value shouldn't change between those calls, right?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java (hunk 9 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: Agreed.

Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]:
```suggestion
            log.debug("For {}, a leader change was detected, currentLeaderEpoch: {}, leaderChangedAttempts: {}",
```

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java [hunk 14]
```diff
@@ -94,9 +100,40 @@ public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, lon
...
+        this.leaderChangedAttempts = -1;
         recordsBuilder.setEstimatedCompressionRatio(compressionRatioEstimation);
     }
 
+    /*
+     * Returns whether the leader epoch has changed since the last attempt.
+     * @param latestLeaderEpoch The latest leader epoch.
+     * @return true if the leader has changed, otherwise false.
+     */
+    boolean hasLeaderChanged(int latestLeaderEpoch) {
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]:
```suggestion
    boolean maybeUpdateLeaderEpoch(int latestLeaderEpoch) {
```

This will imply that we're not just determining that the leader has changed, we're updating internal state.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java [hunk 15]
```diff
@@ -94,9 +100,40 @@ public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, lon
...
+        // Checking for leader change makes sense only from 1st retry onwards(attempt >=1).
+        log.trace("For {}, attempting to change leader, currentLeaderEpoch:{}, leaderChangedAttempts:{}, latestLeaderEpoch: {}, current Attempt: {}",
+            this, currentLeaderEpoch, leaderChangedAttempts, latestLeaderEpoch, attempts());
+        if (attempts() >= 1) {
+            // If the leader's epoch has changed, this counts as a leader change
+            if (currentLeaderEpoch != latestLeaderEpoch) {
+                leaderChangedAttempts = attempts();
+                leaderChanged = true;
+            } else {
+                // Otherwise, it's only a leader change until the first attempt is made with this leader
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: I'm sorry, I don't quite get this logic.

If there are at least one attempt and the above epoch comparison is false, why do we consider the leader as having changed?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java [hunk 16]
```diff
@@ -859,8 +863,9 @@ private List<ProducerBatch> drainBatchesForOneNode(Cluster cluster, Node node, i
                 continue;
 
             Deque<ProducerBatch> deque = getDeque(tp);
-            if (deque == null)
+            if (deque == null) {
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java]: Unless there's a valid reason to, we should omit formatting changes.

File: clients/src/main/java/org/apache/kafka/common/PartitionInfo.java [hunk 17]
```diff
@@ -20,26 +20,35 @@
...
+
+    public PartitionInfo(String topic, int partition, Node leader, Node[] replicas, Node[] inSyncReplicas,
+        Node[] offlineReplicas) {
+        this(topic, partition, leader, UNKNOWN_LEADER_EPOCH, replicas, inSyncReplicas, offlineReplicas);
     }
 
     public PartitionInfo(String topic,
                          int partition,
                          Node leader,
+                         int leaderEpoch,
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/common/PartitionInfo.java]: This is where we might get into trouble since we're breaking a constructor by adding a parameter.

File: clients/src/main/java/org/apache/kafka/common/PartitionInfo.java (hunk 3 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/common/PartitionInfo.java]: `Optional` here, right?

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java (hunk 12 above)
Comment (kirktrue) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java]: I am not personally against logging in tests, but they should be turned down so that they're not emitted by default.

**Review by kirktrue — CHANGES_REQUESTED** [review]
Took a pass, except for tests.

@jsancio - can you weigh in about changes to `org.apache.kafka.common` needing a KIP?

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java (hunk 12 above)
Comment (msn-tldr) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java]: I am inclined to keep it, as it helps improve readability of test logs if/when test fails. 
By default, the logging is turned off in `clients/src/resources/log4.properties`

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java (hunk 11 above)
Comment (msn-tldr) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java]: Inclined to keep it since helps improve readability of test-logs when test fails. By default logging is off, see similar comment below for details.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java (hunk 15 above)
Comment (msn-tldr) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: > least one attempt

implies a retry.

> the above epoch comparison is false

So leader-epoch is still the same. 

Consider leader was changed at attempt=5, to epoch=100. maybeUpdateLeaderEpoch() should detect leader change even when called again at attempt=5, with the same epoch=100. As this is the same attempt in which the leader change was detected.

File: clients/src/main/java/org/apache/kafka/common/PartitionInfo.java [hunk 18]
```diff
@@ -20,6 +20,7 @@
  * This is used to describe per-partition state in the MetadataResponse.
  */
 public class PartitionInfo {
+    public static final int UNKNOWN_LEADER_EPOCH = -1;
```
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/common/PartitionInfo.java]: I think that `PartitionInfo` is a public class so I would not add this here.

Comment (msn-tldr) [reply clients/src/main/java/org/apache/kafka/common/PartitionInfo.java]: @dajac this is removed now, thanks!

**Comment by msn-tldr** [comment]
@kirktrue 
> @jsancio - can you weigh in about changes to org.apache.kafka.common needing a KIP?

I have removed the changes to public classes, so this shouldn't require a KIP, take a look.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java [hunk 19]
```diff
@@ -94,9 +100,45 @@ public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, lon
...
+        this.attemptsWhenLeaderLastChanged = 0;
         recordsBuilder.setEstimatedCompressionRatio(compressionRatioEstimation);
     }
 
+    /*
+     * This is called to update the leader-epoch to which this batch is going to be produced in the ongoing attempt.
+     * @param latestLeaderEpoch The latest leader epoch.
+     * @return true if the leader has changed, otherwise false.
+     */
+    boolean maybeUpdateLeaderEpoch(Optional<Integer> latestLeaderEpoch) {
```
Comment (wcarlson5) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: It seems to be that this doesn't actually update the Leader Epoch, but rather it checks to see if the leader has changed since the last retry. Can we name it something more clear?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java [hunk 20]
```diff
@@ -712,13 +714,14 @@ private long partitionReady(Cluster cluster, long nowMs, String topic,
                 }
 
                 waitedTimeMs = batch.waitedTimeMs(nowMs);
-                backingOff = shouldBackoff(batch, waitedTimeMs);
+                boolean hasLeaderChanged = batch.maybeUpdateLeaderEpoch(leaderAndEpoch.epoch);
```
Comment (wcarlson5) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java]: See that method name and variable name don't really make sense together

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java [hunk 21]
```diff
@@ -1398,16 +1416,253 @@ public void testBuiltInPartitionerFractionalBatches() throws Exception {
...
         }
     }
 
+    /**
+     * For a batch being retried, this validates ready() and drain() whether a batch should skip-backoff(retries-immediately), or backoff, based on -
+     * 1. how long it has waited between retry attempts.
+     * 2. change in leader hosting the partition.
+     */
+    @Test
+    public void testReadyAndDrainWhenABatchIsBeingRetried() throws InterruptedException {
```
Comment (wcarlson5) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java:1432]: I think this is just my unfamiliarity talking, but how are you checking if it is backing off? Is it just relying on these lines? 

```       
assertEquals(Optional.of(part1LeaderEpoch), batch.currentLeaderEpoch());
assertEquals(0, batch.attemptsWhenLeaderLastChanged());
```

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java [hunk 22]
```diff
@@ -1398,16 +1416,253 @@ public void testBuiltInPartitionerFractionalBatches() throws Exception {
...
+            assertEquals(Optional.of(part1LeaderEpoch), batch.currentLeaderEpoch());
+            assertEquals(3, batch.attemptsWhenLeaderLastChanged());
+
+            // Re-enqueue batch for subsequent retries/test-cases.
+            accum.reenqueue(batch, now);
+        }
+    }
+
+    @Test
+    public void testDrainWithANodeThatDoesntHostAnyPartitions() {
```
Comment (wcarlson5) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java:1559]: Is this just added to increase coverage? Thats fine, but I don't see how its related to the rest of the PR

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java (hunk 11 above)
Comment (wcarlson5) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java]: Generally adding logging in tests isn't great. I would second removing it. I don't think the two lines added are worth the change. 

If you want that information about why a batch isn't being processed or if a batch is being processed you should add them as logs to the actual code. They can be trace if necessary

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java [hunk 23]
```diff
@@ -3146,6 +3149,97 @@ public void testInvalidTxnStateIsAnAbortableError() throws Exception {
...
+    public void testProducerBatchRetriesWhenPartitionLeaderChanges() throws Exception {
+        Metrics m = new Metrics();
+        SenderMetricsRegistry senderMetrics = new SenderMetricsRegistry(m);
+        try {
+            // SETUP
+            String metricGrpName = "producer-metrics-test-stats-1";
+            long totalSize = 1024 * 1024;
+            BufferPool pool = new BufferPool(totalSize, batchSize, metrics, time,
+                metricGrpName);
+            long retryBackoffMaxMs = 100L;
```
Comment (wcarlson5) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:3159]: How difficult would it be to mock the backoff using a ticker? Too many times having to rely on timing in tests makes them become flaky. This seems like it would low probability here but we should keep it in mind

**Review by wcarlson5 — COMMENTED** [review]
Overall its looking good. I have a few questions, mostly for my own edification

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java (hunk 15 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: I guess I was under the false impression that `maybeUpdateLeaderEpoch()` was only called _once_ per attempt. I _think_ I can see how the flow of the code would end up calling that multiple times.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java [hunk 24]
```diff
@@ -94,9 +100,45 @@ public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, lon
...
+            if (!currentLeaderEpoch.equals(latestLeaderEpoch)) {
+                attemptsWhenLeaderLastChanged = attempts;
+                leaderChanged = true;
+            } else {
+                // Otherwise, it's only a leader change until the first attempt is made with this leader
+                leaderChanged = attempts == attemptsWhenLeaderLastChanged;
+            }
+        }
+        if (leaderChanged) {
+            log.debug("For {}, leader has changed, oldEpoch: {}, newEpoch: {}",
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: nit: per @wcarlson5's [comment above](https://github.com/apache/kafka/pull/14384/files#r1344314253), the log message might be read that the internal state of the batch was _changed_, rather than simply _detected_. I don't have suggestions on how to reword though. It's fine if you want to leave it as is.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java (hunk 19 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: It does assign the `currentLeaderEpoch` at the end, though, right?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java [hunk 25]
```diff
@@ -682,8 +683,9 @@ private long partitionReady(Cluster cluster, long nowMs, String topic,
             TopicPartition part = new TopicPartition(topic, entry.getKey());
             // Advance queueSizesIndex so that we properly index available
             // partitions.  Do it here so that it's done for all code paths.
-            Node leader = cluster.leaderFor(part);
-            if (leader != null && queueSizes != null) {
+
+            Metadata.LeaderAndEpoch leaderAndEpoch = metadata.currentLeader(part);
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:687]: Since the `leaderAndEpoch` variable isn't reassigned in this method, can we grab a reference to the leader `Node` as we had before? That way we don't have quite so many places that we have the more verbose `leaderAndEpoch.leader.isPresent` and `leaderAndEpoch.leader.get()` everywhere.

Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:687]:
```suggestion
            Metadata.LeaderAndEpoch leaderAndEpoch = metadata.currentLeader(part);
            Node leader = leaderAndEpoch.orElse(null);
```

Or something like that?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java [hunk 26]
```diff
@@ -842,22 +845,31 @@ private boolean shouldStopDrainBatchesForPartition(ProducerBatch first, TopicPar
...
             PartitionInfo part = parts.get(drainIndex);
+
             TopicPartition tp = new TopicPartition(part.topic(), part.partition());
             updateDrainIndex(node.idString(), drainIndex);
             drainIndex = (drainIndex + 1) % parts.size();
             // Only proceed if the partition has no in-flight batches.
             if (isMuted(tp))
                 continue;
-
+            Metadata.LeaderAndEpoch leaderAndEpoch = metadata.currentLeader(new TopicPartition(part.topic(), part.partition()));
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java]: nit: can we reuse `tp` instead of creating a new `TopicPartition`?

**Review by kirktrue — APPROVED** [review]
I have some minor nits, but I'd be OK as is.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java (hunk 15 above)
Comment (msn-tldr) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: > I think I can see how the flow of the code would end up calling that multiple times.

I gets called multiple times. Its sendProducerData -> ready -> partitionReady, and then again in sendProducerData -> drain -> drainBatchesForOneNode.

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java (hunk 11 above)
Comment (msn-tldr) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java]: Ok fair enough will remove the logging from tests.

I have already added traces in actual code, so that should be fine for now.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java (hunk 19 above)
Comment (msn-tldr) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: So the method does 2 things
1. updates leader if a newer leader is supplied
2. also return has changed for the ongoing retry of the batch. 

Keeping them both in 1 method, is a bit confusing, as evident in the naming 😄 

So i have split it into 2 methods, to me code is more readable now.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java (hunk 24 above)
Comment (msn-tldr) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java]: fixed

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java (hunk 21 above)
Comment (msn-tldr) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java:1432]: So if  
1. batch should backoff - then both `ready` & `drain` would not have node1 & batch respectively in their output.
```
assertFalse(result.readyNodes.contains(node1), "Node1 is not ready");
assertTrue(batches.containsKey(node1.id()) && batches.get(node1.id()).isEmpty(),
                "No batches ready to be drained on Node1");
```
2. batch should not be backing off - then both `ready` & `drain` would have node1 & batch respectively in their output.
```
assertTrue(result.readyNodes.contains(node1), "Node1 is ready");
assertTrue(batches.containsKey(node1.id()) && batches.get(node1.id()).size() == 1, "Node1 has 1 batch ready & drained");
```

Other assertions are sanity checking auxillary conditions

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java (hunk 22 above)
Comment (msn-tldr) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java:1559]: `RecordAccumulator` `drain` now uses Metadata Vs Cluster. Metadata can change where as Cluster is immutable. Hence added tests w.r.t metadata changing(which weren't possible earlier)
1. node(broker) hosts partitions but partitions map to it change underneath
2. node(broker) no longer hosts any partitions.

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java (hunk 23 above)
Comment (msn-tldr) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:3159]: Fair point. The `SenderTest` already mocks `time` using `MockTime`, and that is manually advanced. Is that what u had in mind?

Comment (wcarlson5) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:3159]: That's good to know. My concern came from `time.sleep(2 * retryBackoffMaxMs);` down on line 3223. I'm not sure that is avoidable though

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java [hunk 27]
```diff
@@ -94,9 +100,42 @@ public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, lon
...
+            log.trace("For {}, leader wasn't updated, currentLeaderEpoch: {}, attemptsWhenLeaderLastChanged:{}, latestLeaderEpoch: {}, current attempt: {}",
+                this, currentLeaderEpoch, attemptsWhenLeaderLastChanged, latestLeaderEpoch, attempts);
+        }
+    }
+
+    /**
+     * It will return true, for a when batch is being retried, it will be retried to a newer leader.
+     */
+
+    boolean hasLeaderChangedForTheOngoingRetry() {
```
Comment (wcarlson5) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java:128]: love this, thanks!

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java (hunk 22 above)
Comment (wcarlson5) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java:1559]: Awesome, that answer my question perfectly

**Review by wcarlson5 — APPROVED** [review]
Looking good! Thanks for the refactoring bit.

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java (hunk 23 above)
Comment (msn-tldr) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:3159]: `time.sleep(2 * retryBackoffMaxMs);` that is how i ensure that tests waits more than backoff period before retrying.

**Comment by msn-tldr** [comment]
Looking through CI failures - 
https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-14384/8/pipeline/10/

tl;dr - Its flaky test failures, unrelated to this PR.

For JDK8 - There are 2 flaky tests, only 1 tests is related to Java clients `KafkaConsumerTest.testGracefulClose`. I ran this in a loop with Java 1.8, it didn't fail with PR.
https://ge.apache.org/s/255xb6aqxqwuk/tests/overview?outcome=FLAKY,FAILED

For JDK 11 - The 2 flaky runs, and are unrelated to Java clients.
https://ge.apache.org/s/q63xej2ew4w6u

For JDK 17 - Again flaky tests, unrelated to Java client.
https://ge.apache.org/s/dayodofvlspty/tests/overview?outcome=FAILED,FLAKY

For JDK 21 - Flaky tests, unrelated to Java client
https://ge.apache.org/s/cko2ljgyx5oqu/tests/overview?outcome=FLAKY,FAILED

**Comment by msn-tldr** [comment]
@wcarlson5 thanks!

**Comment by ijuma** [comment]
Would it make sense to backport this to 3.6 as well?

**Comment by msn-tldr** [comment]
@ijuma this is merged in 3.6 now, so should be available in 3.6.1.
https://github.com/apache/kafka/pull/14522

**Comment by ijuma** [comment]
Thanks!
